File: ios/TinyTube/BlockHeightStore.swift
import Foundation
import UIKit

/* Remembers what the player measured, so a device that has a usable answer does
   not go asking for another one.

   The counterpart of BlockHeightStore.kt, and it keeps that file's DISPLAY key
   for the same reason: a measurement is about a particular geometry. An iPad in
   a different orientation, Display Zoom being changed, or the same install
   restored onto another device all produce a different key, and a mismatch
   simply measures again.

   IT NO LONGER VERSIONS AND IT NO LONGER GIVES UP. Both were deliberate and
   both were load-bearing, so what removing them costs is written down here
   rather than discovered later:

   - The VERSION key existed because a preference file survives an app update,
     so a wrong answer written by one build is read back by every build after
     it. Bumping it was the only thing that let a corrected measurement reach a
     device that had already stored a bad number. Without it, a stored answer
     outlives any change to how the measurement is made, and a build that
     changes what the measurement returns has no way to invalidate what is
     already on a device short of `clear`.

   - The ATTEMPT COUNT existed because ReplayKit shows a consent alert — once
     per app process, and again after eight minutes backgrounded — and A CHILD
     CAN BE THE ONE LOOKING AT IT. Giving up after three fruitless launches was
     what stopped a device where capture never works from prompting forever.
     Without it, that is exactly what happens: a prompt every launch until some
     capture finally succeeds.

   That second cost is real and current, not hypothetical. It is accepted for
   now because the alternative was worse in the way that matters: the
   measurement is failing on a real device, the blocker is sitting on its
   16-point fallback, and YouTube's seek bar, settings gear, "More videos" and
   logo are all reachable by the child this app exists to fence in. A consent
   alert is recoverable and visible; an unblocked player is the failure the
   whole app is built to prevent.

   Put the attempt count back once the measurement is known to work on real
   hardware, and take the same care with VERSION at the same time. */
enum BlockHeightStore {

    private static let keyPoints = "block_points"
    private static let keyDisplay = "block_display"

    /* Written by builds that versioned and counted attempts. Nothing reads them
       any more; they are named only so `clear` can sweep them off devices that
       still carry them. */
    private static let legacyKeys = ["block_version", "block_sessions"]

    /* What the display is, as a string to compare against later. Points and
       scale rather than raw pixels: the answer is applied in points, so two
       geometries that differ only in a value nothing uses are the same display
       for this purpose. */
    static func displayKey(_ screen: UIScreen = .main) -> String {
        let b = screen.bounds
        return "\(Int(b.width))x\(Int(b.height))@\(screen.scale)"
    }

    /* The remembered height, or nil if there is nothing here worth trusting for
       this display. */
    static func get(_ defaults: UserDefaults = .standard, screen: UIScreen = .main) -> CGFloat? {
        guard defaults.string(forKey: keyDisplay) == displayKey(screen) else { return nil }
        let points = defaults.double(forKey: keyPoints)
        return points > 0 ? CGFloat(points) : nil
    }

    /* Only ever called with a real answer. `Chrome.blockHeight` returns nil for
       "could not tell" precisely so this is never handed a failure dressed up
       as a number. */
    static func put(_ points: CGFloat,
                    defaults: UserDefaults = .standard,
                    screen: UIScreen = .main) {
        defaults.set(Double(points), forKey: keyPoints)
        defaults.set(displayKey(screen), forKey: keyDisplay)
    }

    /* Whether it is worth starting a capture at all — now purely "is there an
       answer for this display yet". No launch budget: a device that has never
       produced one tries again on every restart, for as long as it takes. */
    static func shouldMeasure(_ defaults: UserDefaults = .standard,
                              screen: UIScreen = .main) -> Bool {
        get(defaults, screen: screen) == nil
    }

    /* Throw the answer away and measure again. */
    static func clear(_ defaults: UserDefaults = .standard) {
        for k in [keyPoints, keyDisplay] + legacyKeys {
            defaults.removeObject(forKey: k)
        }
    }
}

File: android/app/src/main/java/dev/vtlinh/tinytube/BlockHeightStore.kt
package dev.vtlinh.tinytube

import android.content.Context

/* Remembers what the player measured off its own paused frame.

   Without this the measurement lived for the life of the process: every cold
   start put the blocker back to the compiled-in fallback and waited for the
   child to pause a video before it was right again — which on a screen with no
   pause button means the first video of every session was played with the
   wrong strip blocked.

   Stored against the display it was measured on. A foldable opening, a change
   to Android's display-size setting, or the same install restored onto another
   phone all change the geometry, and a height measured on the old one would be
   wrong on the new one — so the key is the resolution, and a mismatch simply
   measures again.

   And against a VERSION, which is the more important half. The build that
   introduced this store also, separately, treated a failed capture as an
   answer — so it wrote the fallback here as though it had been measured. A
   preference file survives an app update, so every build afterwards read that
   back, concluded the work was done, and never looked again. Fixing the
   latching bug did nothing on any device that had already run the broken
   build: the wrong answer was on disk.

   So bump VERSION whenever the measurement changes, and old entries are
   ignored rather than trusted. It costs one re-measure per device. */
object BlockHeightStore {

    private const val PREFS = "player"
    private const val KEY_PX = "block_px"
    private const val KEY_DISPLAY = "block_display"
    private const val KEY_VERSION = "block_version"

    /* 1 was never written — the original store had no version key, so anything
       without one came from the build that could persist a failure. 2 was
       written by the builds that measured the RED, which read the scrubber
       knob as the bar's thickness and left the strip a fraction of its proper
       size; 3 measures the track instead and leaves a smaller margin, so
       nothing written before it describes what this build would measure. */
    private const val VERSION = 3

    /* What the display is, as a string to compare against later. */
    fun displayKey(context: Context): String {
        val m = context.resources.displayMetrics
        return "${m.widthPixels}x${m.heightPixels}@${m.densityDpi}"
    }

    private fun prefs(context: Context) =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    /* The remembered height, or null if there is nothing here worth trusting
       for this display and this version of the measurement. */
    fun get(context: Context): Int? {
        val p = prefs(context)
        if (p.getInt(KEY_VERSION, 0) != VERSION) return null
        if (p.getString(KEY_DISPLAY, null) != displayKey(context)) return null
        val px = p.getInt(KEY_PX, -1)
        return if (px >= 0) px else null
    }

    fun put(context: Context, px: Int) {
        prefs(context).edit()
            .putInt(KEY_PX, px)
            .putString(KEY_DISPLAY, displayKey(context))
            .putInt(KEY_VERSION, VERSION)
            .apply()
    }

    /* Throw the answer away and measure again on the next video. */
    fun clear(context: Context) {
        prefs(context).edit()
            .remove(KEY_PX)
            .remove(KEY_DISPLAY)
            .remove(KEY_VERSION)
            .apply()
    }
}
